enterprise/tailnet/pgcoord.go

```diff
@@ -1131,7 +1131,13 @@ func (h *heartbeats) recvBeat(id uuid.UUID) {
 	h.logger.Debug(h.ctx, "got heartbeat", slog.F("other_coordinator_id", id))
 	h.lock.Lock()
 	defer h.lock.Unlock()
-	var oldestTime time.Time
+	if _, ok := h.coordinators[id]; !ok {
+		h.logger.Info(h.ctx, "heartbeats (re)started", slog.F("other_coordinator_id", id))
+		// send on a separate goroutine to avoid holding lock.  Triggering update can be async
+		go func() {
```

Comment:
1. Is it necessary to monitor this goroutine in light of concurrency safety?
2. Is it possible that `sendCtx` is called the second time, but the first execution is still in progress?

Reply:
`sendCtx()` just sends on the given channel, or, if the context expires, returns the context error.  The idea is that if the coordinator is shut down, we don't want any goroutines blocked on sending data across internal channels.  This makes it so we don't have to worry too much about the order in which the different goroutines shut down.  If the channel consumer shuts down first, it's fine because the sender will also close down when the context expires.

So, do directly answer your questions, 

1. No, its not necessary to monitor the goroutine.
2. It is possible that multiple goroutines call `sendCtx()` attempt to write to the channel, but that's explicitly allowed in go.

Reply:
> It is possible that multiple goroutines call sendCtx() attempt to write to the channel

Alright, I assumed that it could be feasible to check that the previous `sendCtx` hangs, so there is no need to spawn more goroutines if we know that there is any congestion.

Reply:
> No, its not necessary to monitor the goroutine.

> if the context expires, returns the context error

To give more context - it happens relatively often in unit tests that a component is leaking a goroutine. I was curious if the coordinator should implement graceful shutdown.

Reply:
> it happens relatively often in unit tests that a component is leaking a goroutine. I was curious if the coordinator should implement graceful shutdown.

Yep, that's why I've implemented these context-limited channel writes, to make it easy to ensure that channel writes are never blocked beyond the given context, even if the receiver goes away.